### PR TITLE
Added support of Onyx based fanout in 'fanouthosts' fixture

### PR DESCRIPTION
**Pull request by yvolynets-mlnx**
### Description of PR

Summary: Added support of Onyx OS based fanout for 'fanouthosts' fixture
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### How did you do it?
Extended 'fanouthosts' fixture with newly created class 'OnyxHost' which transport is based on onyx_config and onyx_command ansible modules.
Also added extended possibility to execute any playbook via fanout instance - for workaround cases when some actions can't be executed via available transport.

#### How did you verify/test it?
Tested on the local setup with PFC Asymmetric test cases.

### CODE REVIEW

File: ansible/group_vars/all/inv_mapping.json [hunk 1]
```diff
@@ -1,4 +1,4 @@
 {
    "lab": "lab_connection_graph.xml",
-   "": "// This is a comment.  specify all internal mappings below this line"
+   "inventory": "lab_connection_graph.xml"
```
Comment (neethajohn) [inline ansible/group_vars/all/inv_mapping.json:3]: I believe this is a comment. can you specify it explicitly?

File: tests/common/fixtures/conn_graph_facts.py [hunk 2]
```diff
@@ -31,3 +31,10 @@ def conn_graph_facts(duthost, testbed_devices):
                 conn_graph_facts = localhost.conn_graph_facts(host=duthost.hostname, filename=lab_conn_graph_file)['ansible_facts']
                 return conn_graph_facts
     return conn_graph_facts
+
+@pytest.fixture(scope="module")
+def fanout_graph_facts(testbed_devices, conn_graph_facts):
+    localhost = testbed_devices["localhost"]
+    fanout_host = conn_graph_facts["device_conn"]["Ethernet0"]["peerdevice"]
+    facts = localhost.conn_graph_facts(host=fanout_host, filename=LAB_CONNECTION_GRAPH)["ansible_facts"]
```
Comment (neethajohn) [inline tests/common/fixtures/conn_graph_facts.py:39]: what is LAB_CONNECTION_GRAPH? Please use the format specified in 'conn_graph_facts' to obtain the correct graph file from the inv mapping so that it can be used by all

File: ansible/group_vars/all/inv_mapping.json (hunk 1 above)
Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.json:3]: Can you please explain more specifically what o you mean?
In json we can't add comments, so I removed empty line to avoid possible errors (if user will specify inventory name not like 'lab').

File: tests/common/fixtures/conn_graph_facts.py (hunk 2 above)
Comment (yvolynets-mlnx) [reply tests/common/fixtures/conn_graph_facts.py:39]: LAB_CONNECTION_GRAPH - missed during merge, it should be direct path to the inventory file.

If to use format like in "conn_graph_facts" fixture, this fixture code should be udpated, to have separate fixture or what that returns current inventory used, then this inventory can be used anywhere. Do you agree?

File: ansible/group_vars/all/inv_mapping.json (hunk 1 above)
Comment (neethajohn) [reply ansible/group_vars/all/inv_mapping.json:3]: The empty string as a key was added to specify the comments to the user on how this inv mapping is supposed to look like (since json doesn't support adding comments)
@yxieca do you have any suggestions on how this should look like. i believe you wanted this to avoid merge conflicts

File: tests/common/fixtures/conn_graph_facts.py (hunk 2 above)
Comment (neethajohn) [reply tests/common/fixtures/conn_graph_facts.py:39]: hardcoding this value will not allow others to make use of this fixture when their connection graph is something other than 'lab_connection_graph.xml'. you need to pass in the correct graph file to 'filename' based on the inventory name the user specifies.

File: ansible/group_vars/all/inv_mapping.json (hunk 1 above)
Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.json:3]: In any case empty string can't be defined here with such value, because code will fail if inventory name will have name other then 'lab', I have tested this and got issue, so thats I removed it.

**Comment by yvolynets-mlnx** [comment]
> 
> 
> This pull request **introduces 1 alert** when merging [a94301d](https://github.com/Azure/sonic-mgmt/commit/a94301d8214318185d34cc7ab07b1bf68b1f01ff) into [f11d076](https://github.com/Azure/sonic-mgmt/commit/f11d076cadc638a64270f31e8b68b17ba403c6b4) - [view on LGTM.com](https://lgtm.com/projects/g/Azure/sonic-mgmt/rev/pr-c2c00cd56af128a2306c4698cc0182aae40fbcc2)
> 
> **new alerts:**
> 
>     * 1 for Module is imported more than once

I suggest to skip this alert.

File: ansible/group_vars/all/inv_mapping.json (hunk 1 above)
Comment (wangxin) [reply ansible/group_vars/all/inv_mapping.json:3]: Maybe we can consider converting this inv_mapping.json to yaml format. Yaml file supports comments.

Comment (neethajohn) [reply ansible/group_vars/all/inv_mapping.json:3]: If inventory name is something other than 'lab', it needs to be defined here along with its corresponding graph file. 
I suggest adding a comment line as follows 
"your_inventory_name" : "graph file for your inventory // sample inv graph mapping. Please specify all internal mappings below this line"
If you want to go the yaml route, I am okay with that too.

File: tests/common/devices.py [hunk 3]
```diff
@@ -663,6 +736,9 @@ def __init__(self, ansible_adhoc, os, hostname, device_type, user, passwd):
         if os == 'sonic':
             self.os = os
             self.host = SonicHost(ansible_adhoc, hostname)
+        elif os == 'onyx':
+            self.os = os
+            self.host = OnyxHost(ansible_adhoc, hostname, user, passwd)
```
[the remark this answers is not in the document]
Comment (yvolynets-mlnx) [reply tests/common/devices.py:741]: It is already handled in PFC Assym PR
https://github.com/Azure/sonic-mgmt/blob/307469cc9ed93830d63da3f4fc5a73e6af4c190e/tests/common/fixtures/pfc_asym.py#L75

File: ansible/group_vars/all/inv_mapping.json (hunk 1 above)
Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.json:3]: In such case, lets move from json to yml file. I will add this change next week.

Comment (yxieca) [reply ansible/group_vars/all/inv_mapping.json:3]: I am ok with moving to yml. The 'comments' is to create enough gap so that github change won't always request manual merge when rebasing internal repos.

Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.json]: @neethajohn @yxieca @wangxin Please take a look recent changes, thanks.

File: tests/common/fixtures/conn_graph_facts.py (hunk 2 above)
Comment (yvolynets-mlnx) [reply tests/common/fixtures/conn_graph_facts.py]: Fixed

File: ansible/group_vars/all/inv_mapping.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+# Specify all internal mappings below
```
Comment (yxieca) [inline ansible/group_vars/all/inv_mapping.yml:2]: next 2 lines are actually public changes. Please fix wording.

File: ansible/group_vars/all/inv_mapping.yml [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+lab: lab_connection_graph.xml
+inventory: lab_connection_graph.xml
```
Comment (yxieca) [inline ansible/group_vars/all/inv_mapping.yml:4]: Please add 3 line gap after inventory with comments. That way creates nice separation for any internal changes after the gap.

Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.yml:4]: Fixed

File: ansible/group_vars/all/inv_mapping.yml (hunk 4 above)
Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.yml]: Fixed

**Comment by yvolynets-mlnx** [comment]
> 
> 
> > This pull request **introduces 1 alert** when merging [a94301d](https://github.com/Azure/sonic-mgmt/commit/a94301d8214318185d34cc7ab07b1bf68b1f01ff) into [f11d076](https://github.com/Azure/sonic-mgmt/commit/f11d076cadc638a64270f31e8b68b17ba403c6b4) - [view on LGTM.com](https://lgtm.com/projects/g/Azure/sonic-mgmt/rev/pr-c2c00cd56af128a2306c4698cc0182aae40fbcc2)
> > **new alerts:**
> > ```
> > * 1 for Module is imported more than once
> > ```
> 
> Will be fixed.

Fixed

File: tests/common/devices.py [hunk 6]
```diff
@@ -647,6 +649,79 @@ def check_bgp_session_state(self, neigh_ips, neigh_desc, state="established"):
 
         return False
 
+    def exec_template(self, ansible_root, ansible_playbook, **kwargs):
+        playbook_template = 'cd {ansible_path}; ansible-playbook {playbook} -i lab -l {fanout_host} --extra-vars \'{extra_vars}\' -vvvvv'
```
Comment (neethajohn) [inline tests/common/devices.py:653]: Do not hardcode inventory file name here

File: tests/common/devices.py [hunk 7]
```diff
@@ -647,6 +649,79 @@ def check_bgp_session_state(self, neigh_ips, neigh_desc, state="established"):
...
+            lines=['lacp rate %s' % mode],
+            parents='interface ethernet %s' % interface_name)
+        logging.info("Set interface [%s] lacp rate to [%s]" % (interface_name, mode))
+        return out
+
+    def exec_template(self, ansible_root, ansible_playbook, **kwargs):
+        """
+        Execute ansible playbook with specified parameters
+        """
+        playbook_template = 'cd {ansible_path}; ansible-playbook {playbook} -i lab -l {fanout_host} --extra-vars \'{extra_vars}\' -vvvvv'
```
Comment (neethajohn) [inline tests/common/devices.py:717]: Do not hardcode inventory filename

File: tests/common/fixtures/conn_graph_facts.py [hunk 8]
```diff
@@ -1,17 +1,33 @@
...
     base_path = os.path.dirname(os.path.realpath(__file__))
-    # json file contains mapping from inventory file name to its corresponding graph file
-    inv_mapping_file = os.path.join(base_path, "../../../ansible/group_vars/all/inv_mapping.json")
+    # yaml file contains mapping from inventory file name to its corresponding graph file
+    inv_mapping_file = os.path.join(base_path, "../../../ansible/group_vars/all/inv_mapping.yml")
     if os.path.exists(inv_mapping_file):
         with open(inv_mapping_file) as fd:
-            inv_map = json.load(fd)
+            inv_map = yaml.load(fd, Loader=yaml.FullLoader)
+        import pdb; pdb.set_trace()
```
Comment (neethajohn) [inline tests/common/fixtures/conn_graph_facts.py:30]: remove

File: tests/conftest.py [hunk 9]
```diff
@@ -97,6 +97,11 @@ def pytest_addoption(parser):
     parser.addoption("--logs_since", action="store", type=int,
                     help="number of minutes for show techsupport command")
 
+    # fw_utility options
+    parser.addoption("--config_file", action="store", default=None, help="name of configuration file (per each vendor)")
+    parser.addoption("--binaries_path", action="store", default=None, help="path to binaries files")
+    parser.addoption("--second_image_path", action="store", default=None, help="path to second image to be installed if there is no image available")
```
Comment (neethajohn) [inline tests/conftest.py:103]: What are these changes? They don't seem to be related to the fanout fixture

File: tests/common/fixtures/conn_graph_facts.py [hunk 10]
```diff
@@ -1,17 +1,33 @@
...
 @pytest.fixture(scope="module")
 def conn_graph_facts(duthost, localhost):
-    conn_graph_facts = dict()
+    conn_graph_facts = get_graph_facts(duthost, localhost, duthost.hostname)
+    return conn_graph_facts
+
+  
+@pytest.fixture(scope="module")
+def fanout_graph_facts(localhost, duthost, conn_graph_facts):
+    fanout_host = conn_graph_facts["device_conn"]["Ethernet0"]["peerdevice"]
```
Comment (neethajohn) [inline tests/common/fixtures/conn_graph_facts.py:13]: We have setups where a DUT is connected to multiple fanouts.

File: ansible/group_vars/all/inv_mapping.yml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+---
+# Specify all public mappings below
+lab: lab_connection_graph.xml
+inventory: lab_connection_graph.xml
```
Comment (yxieca) [inline ansible/group_vars/all/inv_mapping.yml:4]: Sorry missed this one. Why do we need to have inventory file here?

@neethajohn , @wangxin comments?

File: tests/common/fixtures/conn_graph_facts.py (hunk 10 above)
Comment (yvolynets-mlnx) [reply tests/common/fixtures/conn_graph_facts.py:13]: I believe that this is out of scope of this PR, because main goal of this changes was adding support of Onyx based fanout. I would like to have it merged first and the we can take multiple fanout support as another PR since we don't have such configurations and I'm not able to verify it.

File: ansible/group_vars/all/inv_mapping.yml (hunk 11 above)
Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.yml:4]: @neethajohn @wangxin could you please reply to this comment?

File: tests/common/devices.py (hunk 6 above)
Comment (yvolynets-mlnx) [reply tests/common/devices.py]: Fixed.

File: tests/common/fixtures/conn_graph_facts.py (hunk 8 above)
Comment (yvolynets-mlnx) [reply tests/common/fixtures/conn_graph_facts.py]: fixed

File: tests/conftest.py (hunk 9 above)
Comment (yvolynets-mlnx) [reply tests/conftest.py]: Fixed

File: tests/common/devices.py (hunk 7 above)
Comment (yvolynets-mlnx) [reply tests/common/devices.py]: Fixed.

File: ansible/group_vars/all/inv_mapping.yml (hunk 11 above)
Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.yml:4]: @neethajohn @wangxin could you please reply to this comment?

**Comment by yvolynets-mlnx** [comment]
@neethajohn @wangxin Would be nice to checkthis PR and merge, because https://github.com/Azure/sonic-mgmt/pull/1534 is blocked

File: ansible/group_vars/all/inv_mapping.yml (hunk 11 above)
Comment (neethajohn) [reply ansible/group_vars/all/inv_mapping.yml:4]: @yvolynets-mlnx , I guess Ying's question is why there is a key named 'inventory' in that file mapping to 'lab_connection_graph.xml'.

File: tests/common/fixtures/conn_graph_facts.py (hunk 10 above)
Comment (neethajohn) [reply tests/common/fixtures/conn_graph_facts.py:13]: Sure. I can add in those changes

Comment (yvolynets-mlnx) [reply tests/common/fixtures/conn_graph_facts.py:13]: Thank you

File: ansible/group_vars/all/inv_mapping.yml (hunk 11 above)
Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.yml:4]: Different inventory file names can have the same graph file mapping.
So "inventory" file also was added here, as it is one of main inventory files.

Comment (yxieca) [reply ansible/group_vars/all/inv_mapping.yml:4]: Yuri, I don't think we have an inventory called 'inventory' unless it is something Mellanox specific? If so, can you add it in your internal repo instead?

I am ok with merging this PR first and continue discussion and make fix/updates later because I believe this change won't break anything.

Comment (yvolynets-mlnx) [reply ansible/group_vars/all/inv_mapping.yml:4]: Yes we can add this "inventory" key internally. However there is the same file in master as well, which is updated to our needs internally - https://github.com/Azure/sonic-mgmt/blob/master/ansible/inventory.
I suppose some other user also could use this file as it does exist for a long time.

**Comment by qiluo-msft** [comment]
Could you fix the LGTM new alert?
